k1te-backend/src/main/java/ua/com/pragmasoft/k1te/backend/router/domain/KiteRouter.java

```diff
@@ -59,13 +64,44 @@ public void dispatch(RoutingContext ctx) throws KiteException {
     if (null == ctx.destinationConnection) {
       ctx.destinationConnection = ctx.to.getConnectionUri();
     }
-    Connector connector = requiredConnector(this.connectorId(ctx.destinationConnection));
+    Connector connector = requiredConnector(Connector.connectorId(ctx.destinationConnection));
     connector.dispatch(ctx);
-    if (null == ctx.response) {
+    MessageAck response = ctx.response;
+    if (null == response) {
       throw new RoutingException("missing response from connector " + connector.id());
     }
-    this.channels.updatePeer(ctx.to, ctx.from.getId());
-    this.channels.updatePeer(ctx.from, ctx.to.getId());
+
+    String messageId = response.messageId();
+    String destinationMessageId = response.destinationMessageId();
+    if (!ctx.isIdle) { // if isIdle is true does nothing
+      if (!messageId.equals("-")) { // do nothing if it's join/left/switch/selfMessage messages
+        String ownerMessageId;
+        String toMessageId;
+        if (Connector.connectorId(ctx.originConnection).equals(WsConnector.WS)) {
+          ownerMessageId = destinationMessageId;
+          toMessageId = destinationMessageId;
+        } else if (Connector.connectorId(ctx.destinationConnection).equals(WsConnector.WS)) {
+          ownerMessageId = messageId;
+          toMessageId = messageId;
+        } else {
+          ownerMessageId = messageId;
+          toMessageId = destinationMessageId;
+        }
+
+        this.channels.updateUri(ctx.from, ctx.originConnection, ownerMessageId, Instant.now());
+        this.channels.updateUri(ctx.to, ctx.destinationConnection, toMessageId, Instant.now());
+
+        String content = HistoryMessage.encode(ctx.request);
```

Comment:
I thought about moving things like history updater to separate routing listeners (sync, vetoable) or pre/post processors, to allow for more extensibility and reduce coupling.. For now it's probably bearable but next time we will have to add logic to the router, lets redesign the routing flow..

Reply:
Agree, I think PostProcessor will be good solution here, it will allow us to complement Router logic without making it more arduous.